File: 백준/Gold/2636. 치즈/치즈.py
from collections import deque

dx = [0,0,-1,1]
dy = [1,-1,0,0]

def bfs():
    cnt=0
    dq = deque([[0,0]])
    visited[0][0]=1

    while dq:
        x,y = dq.popleft()
        for i in range(4):
            xx = x + dx[i]
            yy = y + dy[i]
            if 0<=xx<n and 0<=yy<m and visited[xx][yy]==0:
                visited[xx][yy]=1
                if arr[xx][yy]==0:
                    dq.append([xx,yy])
                else:
                    arr[xx][yy]=0
                    cnt+=1
    return cnt


n,m = map(int,input().split())
arr = [list(map(int,input().split())) for _ in range(n)]
res=[]
time = 0
while True:
    time += 1
    visited = [[0]*m for _ in range(n)]
    cnt = bfs()
    res.append(cnt)

    if cnt==0:
        break

print(time-1)
print(res[-2])

File: 백준/Gold/2636. 치즈/치즈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int n,m;
	static int[][] arr;
	static boolean[][] visited;
	static Queue<int[]> dq = new ArrayDeque<int[]>();
	static int[] dx = {-1,1,0,0};
	static int[] dy = {0,0,-1,1};
	
	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		
		visited = new boolean[n][m];
		arr = new int[n][m];
		for(int i=0;i<n;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<m;j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		dq.add(new int[] {0,0});
		visited[0][0] = true;
		int time=0;
		int precnt=0;
		while(!dq.isEmpty()) {
			Queue<int[]> tmpdq = new ArrayDeque<int[]>();
			int cnt=0;
			while(!dq.isEmpty()) {
				int[] xy = dq.poll();
				
				
				for(int i=0;i<4;i++) {
					int x = xy[0] + dx[i];
					int y = xy[1] + dy[i];
					if((x>=0)&&(y>=0)&&(x<n)&&(y<m)&&(!visited[x][y])) {
						visited[x][y]=true;
						if(arr[x][y]==0) {
							dq.add(new int[] {x,y});
						}else {
							cnt+=1;
							tmpdq.add(new int[] {x,y});
						}
					}
				}
			}
			
			if(cnt == 0)
				break;
			
			precnt=cnt;
			time++;

			dq=tmpdq;
			
		
		}

		
		System.out.println(time);
		System.out.println(precnt);
		
	}

}